src/nupic/math/Utils.hpp

```diff
@@ -352,33 +351,59 @@ private:                                                                       \
           for (UInt m = 0; m < R; ++m)                                         \
             for (UInt n = 0; n < S; ++n)
 
-/**
- * Function object that takes a single argument, a pair (or at least
- * a class with the same interface as pair), and returns the pair's
- * first element. This is not part of the C++ standard, but usually
- * provided by implementations of STL.
- */
-template <class Pair>
-struct select1st : public std::unary_function<Pair, typename Pair::first_type> {
-  inline const typename Pair::first_type &operator()(const Pair &x) const {
-    return x.first;
-  }
-};
+  /**
+   * This is a 'C++ function object' or Functor.  An object that can be passed
+   * as if it were a C function. It is created by having a class containing an
+   * overload of the () operator.
+   *
+   * Function object that takes a single argument, a pair (or at least
+   * a class with the same interface as pair), and returns the pair's
+   * first argument.
+   */
+#if __cplusplus >= 201402L || (defined(_MSC_VER) && _MSC_VER >= 1914)
+  // C++14 or higher
```

Comment:
go to c++14 and that's it..?

Reply:
unary_function is not valid in C++17.  But its replacement; an auto that can handle a ( ) overload, is not available until C++14.  So to do both C++11 and C++17 we need a conditional compile.